### PR TITLE
feat(autosetup)!: Refactor autosetup process

### PR DESCRIPTION
**Pull request by Phil91**
The autosetup process is changed for offers of type Service 
For ServiceType DATASPACE_SERVICE only the subscription record and if set a technical user is created
For ServiceType CONSULTANCE_SERVICE only the subscription record is created

Refs: CPLP-1845

### CODE REVIEW

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 1]
```diff
@@ -67,4 +103,170 @@ public async Task AutoSetupOffer(OfferThirdPartyAutoSetupData autoSetupData, str
...
+            subscription.OfferSubscriptionStatusId = OfferSubscriptionStatusId.ACTIVE;
+            subscription.LastEditorId = offerDetails.CompanyUserId;
+        });
+
+        var userRolesRepository = _portalRepositories.GetInstance<IUserRolesRepository>();
+        var technicalUserClientId = $"{offerDetails.OfferName}-{offerDetails.CompanyName}";
+        ClientInfoData? clientInfoData = null;
+        if (offerTypeId == OfferTypeId.APP)
+        {
+            var (clientId, iamClientId) = await CreateClient(data, userRolesRepository, offerDetails);
```
Comment (ntruchsess) [inline src/marketplace/Offers.Library/Service/OfferSetupService.cs:122]: consider validation of data.OfferUrl to be wellformed ( I suggest use of https://github.com/catenax-ng/tx-portal-backend/blob/705cedf1b8b2a97441e89bc71abf8eac1927e137/src/framework/Framework.IO/UrlHelper.cs#L30 )

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 2]
```diff
@@ -67,4 +103,170 @@ public async Task AutoSetupOffer(OfferThirdPartyAutoSetupData autoSetupData, str
...
+                nameof(offerDetails.CompanyUserId));
+        }
+
+        return offerDetails;
+    }
+
+    private async Task<(string clientId, Guid iamClientId)> CreateClient(OfferAutoSetupData data, IUserRolesRepository userRolesRepository, OfferSubscriptionTransferData offerDetails)
+    {
+        var userRoles = await userRolesRepository.GetUserRolesForOfferIdAsync(offerDetails.OfferId).ConfigureAwait(false);
+        var redirectUrl = data.OfferUrl.EndsWith("/") ? $"{data.OfferUrl}*" : $"{data.OfferUrl}/*";
```
Comment (ntruchsess) [inline src/marketplace/Offers.Library/Service/OfferSetupService.cs:176]: I suggest to use the existing framework-method to append to path: https://github.com/catenax-ng/tx-portal-backend/blob/705cedf1b8b2a97441e89bc71abf8eac1927e137/src/framework/Framework.IO/UrlHelper.cs#L56

File: src/portalbackend/PortalBackend.PortalEntities/Entities/ServiceDetail.cs [hunk 3]
```diff
@@ -22,20 +22,25 @@
...
         ServiceId = serviceId;
         ServiceTypeId = serviceTypeId;
+        TechnicalUserNeeded = technicalUserNeeded;
     }
 
     public Guid ServiceId { get; private set; }
+
     public ServiceTypeId ServiceTypeId { get; private set; }
 
+    public bool TechnicalUserNeeded { get; private set; }
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/Entities/ServiceDetail.cs:40]: shouldn't this be settable to support later update? We only define columns as private set if there is a good reason (like a foreign key for a relationship that must never change (e.g. ServiceDetails being a Service-specific extension of Offer for a specific service) will never be reassigned to a different offer)

File: src/portalbackend/PortalBackend.PortalEntities/PortalDbContext.cs [hunk 4]
```diff
@@ -1080,5 +1062,20 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
+        {
+            entity.HasKey(e => new {e.ServiceId, e.ServiceTypeId});
+
+            entity.HasOne(e => e.ServiceType)
+                .WithMany(e => e.ServiceDetails)
+                .HasForeignKey(e => e.ServiceTypeId)
+                .OnDelete(DeleteBehavior.ClientSetNull);
+
+            entity.HasOne(e => e.Service)
+                .WithMany(e => e.ServiceDetails)
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/PortalDbContext.cs:1076]: is this really 1-N? Shall a specific Service really have multiple sets of service-specific properties assigned? (like can a service belong to multiple ServiceTypes? and if so does the same apply to all other detail properties like the requirement to create a technical user?)

File: src/provisioning/Provisioning.Library/Extensions/ServiceAccountManager.cs [hunk 5]
```diff
@@ -27,8 +27,11 @@ namespace Org.Eclipse.TractusX.Portal.Backend.Provisioning.Library;
 
 public partial class ProvisioningManager
 {
-    public async Task<string> GetNextServiceAccountClientIdAsync() =>
-        _Settings.ServiceAccountClientPrefix + (await _ProvisioningDBAccess!.GetNextClientSequenceAsync().ConfigureAwait(false));
+    public async Task<(string clientId, string id)> GetNextServiceAccountClientIdWithIdAsync()
+    {
+        var id = await _ProvisioningDBAccess!.GetNextClientSequenceAsync().ConfigureAwait(false);
+        return ($"{_Settings.ServiceAccountClientPrefix}{id}", id.ToString());
```
Comment (ntruchsess) [inline src/provisioning/Provisioning.Library/Extensions/ServiceAccountManager.cs:33]: As there is only a single reference to this method consider relocating the complete method from ProvisioningManager into ServiceAccountCreation-service.

**Review by ntruchsess — CHANGES_REQUESTED** [review]
I did minor cleanup - please pull before working on the comments

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs (hunk 1 above)
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs:100]: check added

File: src/portalbackend/PortalBackend.PortalEntities/PortalDbContext.cs (hunk 4 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/PortalDbContext.cs:1076]: yes, it's possible that one service has multiple ServiceTypes and e.g. only one of them has a technical user

File: src/provisioning/Provisioning.Library/Extensions/ServiceAccountManager.cs (hunk 5 above)
Comment (Phil91) [reply src/provisioning/Provisioning.Library/Extensions/ServiceAccountManager.cs]: moved to ServiceAccountCreationService

File: src/portalbackend/PortalBackend.PortalEntities/Entities/ServiceDetail.cs (hunk 3 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/ServiceDetail.cs]: absolutely right, removed the private of the setter

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs (hunk 2 above)
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs]: good advice. adjusted

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 6]
```diff
@@ -20,28 +20,64 @@
...
+    public async Task AutoSetupOfferSubscription(OfferThirdPartyAutoSetupData autoSetupData, string accessToken, string autoSetupUrl)
     {
-        _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+        var httpClient = _httpClientFactory.CreateClient(nameof(OfferSetupService));
+        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
 
         try
         {
-            var response = await _httpClient.PostAsJsonAsync(serviceDetailsAutoSetupUrl, autoSetupData).ConfigureAwait(false);
+            var response = await httpClient.PostAsJsonAsync(autoSetupUrl, autoSetupData).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs]: good advice. adjusted

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**